nsv13/code/_onclick/drag_drop.dm

```diff
@@ -0,0 +1,63 @@
+/obj/item/gun
+	var/atom/autofire_target = null //What are we aiming at? This will change if you move your mouse whilst spraying.
+
+/obj/item/gun/proc/can_fire_at(atom/target, mob/user)
+	if(!target)
+		return FALSE
+
+	if(target == user) //so we can't shoot ourselves with autofire
+		return FALSE
+
+	if(user.get_active_held_item() != src)
+		return FALSE
+
+	if(istype(target, /obj/screen))
+		return FALSE
+
+	if(target in user.contents) //can't shoot stuff inside us.
+		return FALSE
+
+	if(user.a_intent == INTENT_HARM) //melee attacks are handled by attackby, not autofire
+		return FALSE
+
+	if(isliving(user))//Check if the user can use the gun, if the user isn't alive(turrets) assume it can.
+		var/mob/living/L = user
+		if(!can_trigger_gun(L))
+			return FALSE
+		if(clumsy_check)//If theyre a clown, they can't autofire.
+			if (HAS_TRAIT(user, TRAIT_CLUMSY))
+				return FALSE
+
+	if(!can_shoot()) //Just because you can pull the trigger doesn't mean it can shoot. (*CLICK*)
+		return FALSE
+
+	if(weapon_weight == WEAPON_HEAVY && user.get_inactive_held_item()) //If you can't fire it fully, you can't autofire it.
+		return FALSE
+
+	return TRUE //If we passed all those other checks. Yay, we get to fire. Hoorah.
+
+/obj/item/gun/ballistic/automatic/burst_select()
+	. = ..()
+	autofire_target = null
+
+/obj/item/gun/onMouseUp(object, location, params, mob)
+	autofire_target = null
+	return ..()
+
+/obj/item/gun/onMouseDown(object, location, params)
+	set waitfor = FALSE //Asynchronous processing is required here due to the while loop. In other words. Don't hold up the client's clicking while we run a loop that can go on for ages (potentially!)
+	var/mob/user = src.loc
+	autofire_target = object //When we start firing, we start firing at whatever you clicked on initially. When the user drags their mouse, this shall change.
+	while(autofire_target)  //While will only run while we have a user (loc) that is a mob, and we are being actively held by this mob, they have a client (as to prevent disconnecting mid-fight causing you to perma-fire) and of course, if we passed the previous check about autofiring.
+		stoplag(max(fire_delay, 0.15 SECONDS)) //Default fire delay to prevent you from instantly dumping an entire mag out.
```

Comment:
Is the default fire delay a DEFINE? If not I suggest making it one as to properly prevent issues by balance changes down the line.

Reply:
Default is 0.